### PR TITLE
C# 7.x: Add support for readonly structs

### PR DESCRIPTION
**Pull request by RexJaeschke**
This PR adds one new modifier to structs, `readonly`, in *struct_modifier*.

### CODE REVIEW

**Comment by KalleOlaviNiemitalo** [comment]
Should an implicit conversion from `dynamic` to a ref struct type be rejected at compile time? Microsoft's implementation of `dynamic` uses the ref struct as a type argument, which is then not supported by the CLR. However, as the C# standard does not specify the details of `dynamic`, I imagine a different compiler with a different library could find a way to make the conversion work. <https://github.com/dotnet/csharplang/discussions/4713>, <https://github.com/dotnet/csharpstandard/issues/282>

File: standard/structs.md [hunk 1]
```diff
@@ -38,12 +38,38 @@ struct_modifier
...
+The `ref` modifier indicates that the *struct_declaration* declares a type whose instances are allocated on the execution stack.
+
+It is a compile-time error if a ref struct type is used in any of the following contexts:
+- As the element type of an array.
+- As the declared type of a field of a class or a non-ref struct.
+- To implement an interface.
+- Being boxed to `System.ValueType` or `System.Object`.
+- As a type argument.
+- An async method.
+- An iterator.
```
Comment (KalleOlaviNiemitalo) [inline standard/structs.md]: Does "a ref struct type is used" include this case:

```C#
using System;
using System.Threading.Tasks;

class Widget
{
    public static implicit operator Widget(Span<int> s) => null;
}

static class Class1
{
    static async Task M(Task t)
    {
        Widget w = stackalloc int[] { 42 };
        await t;
    }
}
```

.NET SDK 6.0.101 compiles it without errors. The Span\<int\> becomes a local variable in the MoveNext method of the state machine.

**Comment by gafter** [comment]
This draft does not reflect the draft specification at https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md

**Comment by RexJaeschke** [comment]
Reminder to look at https://github.com/dotnet/csharpstandard/issues/334 for other open issues w.r.t this PR (including Neal's concerns re missing text re span safety).

**Comment by RexJaeschke** [comment]
Based on Neal’s suggestion during yesterday's TG2 meeting, we agreed to merge as is, but only call `readonly structs` completed. Then, create another PR for all the additional work for `ref structs`.

**Comment by BillWagner** [comment]
I'll rebase this branch to resolve conflicts, and mark it ready for review.

**Comment by jskeet** [comment]
Assigned to @gafter for the splitting off - we should have assigned this before...

**Comment by gafter** [comment]
I've removed ref struct from this PR and moved it to https://github.com/dotnet/csharpstandard/pull/601.  This should be fine for `readonly struct`.

File: standard/structs.md (hunk 1 above)
Comment (BillWagner) [reply standard/structs.md]: Resolving this conversation in this PR, as this concern should be covered in #601

Comment (KalleOlaviNiemitalo) [reply standard/structs.md]: Reposted as <https://github.com/dotnet/csharpstandard/pull/601#discussion_r925709489>.

**Comment by BillWagner** [comment]
@jskeet 

We should be able to do the final approval for this and merge it during this week's meeting. The `ref` struct portion has all been moved to #601

**Review by BillWagner — APPROVED** [review]
This LGTM. Let's :shipit:

**Review by jskeet — APPROVED** [review]

**Comment by jskeet** [comment]
Yup, just done a final review. Will merge now.

**Pull request merged**